lib/web_ui/lib/src/engine/canvaskit/surface.dart

```diff
@@ -171,11 +170,49 @@ class Surface {
       ..height = '${logicalHeight}px';
   }
 
+  void _contextRestoredListener(html.Event event) {
+    print('Flutter: restoring WebGL context.');
+    assert(
+        _contextLost,
+        'Received "webglcontextrestored" event but never received '
+        'a "webglcontextlost" event.');
+    _contextLost = false;
+    // Force the framework to rerender the frame.
+    EnginePlatformDispatcher.instance.invokeOnMetricsChanged();
+    event.stopPropagation();
+    event.preventDefault();
+  }
+
+  void _contextLostListener(html.Event event) {
+    if (event.target != this.htmlCanvas) {
+      // This is the context lost event for the old HTML canvas. Ignore it.
+      return;
+    }
+    final SurfaceFactory factory = SurfaceFactory.instance;
+    _contextLost = true;
+    if (factory.isLive(this)) {
+      _forceNewContext = true;
+      event.preventDefault();
+    } else {
+      dispose();
+    }
+  }
+
   /// This function is expensive.
   ///
   /// It's better to reuse surface if possible.
   CkSurface _createNewSurface(ui.Size physicalSize) {
     // Clear the container, if it's not empty. We're going to create a new <canvas>.
+    this.htmlCanvas?.removeEventListener(
+          'webglcontextrestored',
+          _contextRestoredListener,
```

Comment:
This may not be correctly removed. We've had similar issues before (e.g. [here](https://github.com/flutter/engine/blob/8a562b906d66085dc53bd7f0506ee557ab0e7904/lib/web_ui/lib/src/engine/keyboard.dart#L36-L67)).

My guess is that this is happening because we are getting a new function from the tear off. So we end up adding a function as a listener, then trying to remove a different function.

Reply:
If we do the removal correctly, we don't need to worry about the target anymore.

Reply:
Thanks! This was perplexing me